File: client-js/src/http/HttpClient.ts
/**
 * HTTP client with retry, load balancing, and failover support
 */

import * as logger from '../utils/logger'
import type { HttpClientOptions } from '../types'
import type { LoadBalancer } from './LoadBalancer'

export class HttpClient {
  private baseUrl: string | null
  private loadBalancer: LoadBalancer | null
  private timeoutMillis: number
  private retryAttempts: number
  private retryDelayMillis: number
  private enableFailover: boolean
  private bearerToken: string | null
  private headers: Record<string, string>

  constructor(options: HttpClientOptions = {}) {
    const {
      baseUrl = null,
      loadBalancer = null,
      timeoutMillis = 30000,
      retryAttempts = 3,
      retryDelayMillis = 1000,
      enableFailover = true,
      bearerToken = null,
      headers = {}
    } = options

    this.baseUrl = baseUrl ?? null
    this.loadBalancer = loadBalancer ?? null
    this.timeoutMillis = timeoutMillis
    this.retryAttempts = retryAttempts
    this.retryDelayMillis = retryDelayMillis
    this.enableFailover = enableFailover
    this.bearerToken = bearerToken ?? null
    this.headers = headers || {}

    logger.log('HttpClient.constructor', {
      hasLoadBalancer: !!loadBalancer,
      baseUrl: baseUrl || 'load-balanced',
      timeoutMillis,
      retryAttempts,
      enableFailover,
      hasAuth: !!bearerToken,
      customHeaders: Object.keys(this.headers).length
    })
  }

  private async executeRequest(url: string, method: string, body: unknown = null, requestTimeoutMillis: number | null = null): Promise<unknown> {
    const effectiveTimeout = requestTimeoutMillis || this.timeoutMillis
    logger.log('HttpClient.request', { method, url, hasBody: !!body, timeout: effectiveTimeout })

    const controller = new AbortController()
    const timeoutId = setTimeout(() => controller.abort(), effectiveTimeout)

    try {
      const headers: Record<string, string> = { 'Content-Type': 'application/json' }
      if (this.bearerToken) {
        headers['Authorization'] = `Bearer ${this.bearerToken}`
      }
      Object.assign(headers, this.headers)

      const options: RequestInit = {
        method,
        signal: controller.signal,
        headers
      }

      if (body) {
        options.body = JSON.stringify(body)
      }

      const response = await fetch(url, options)

      logger.log('HttpClient.response', { method, url, status: response.status })

      if (response.status === 204) {
        return null
      }

      if (!response.ok) {
        const error: Error & { status?: number } = new Error(`HTTP ${response.status}: ${response.statusText}`)
        error.status = response.status

        try {
          const text = await response.text()
          if (text) {
            const body = JSON.parse(text)
            error.message = body.error || error.message
          }
        } catch (_e) {
          // Ignore JSON parse errors
        }

        logger.error('HttpClient.request', { method, url, status: error.status, error: error.message })
        throw error
      }

      const contentType = response.headers.get('content-type')
      const contentLength = response.headers.get('content-length')

      if (!contentType || !contentType.includes('application/json') || contentLength === '0') {
        const text = await response.text()
        if (!text || text.length === 0) {
          return null
        }
        try {
          return JSON.parse(text)
        } catch (_e) {
          return null
        }
      }

      return response.json()

    } catch (error) {
      if (error instanceof Error && error.name === 'AbortError') {
        const timeoutError: Error & { timeout?: number } = new Error(`Request timeout after ${effectiveTimeout}ms`)
        timeoutError.name = 'AbortError'
        timeoutError.timeout = effectiveTimeout
        logger.error('HttpClient.request', { method, url, error: 'timeout', timeout: effectiveTimeout })
        throw timeoutError
      }
      logger.error('HttpClient.request', { method, url, error: (error as Error).message })
      throw error
    } finally {
      clearTimeout(timeoutId)
    }
  }

  private async requestWithRetry(method: string, path: string, body: unknown = null, requestTimeoutMillis: number | null = null): Promise<unknown> {
    let lastError: Error | null = null

    for (let attempt = 0; attempt < this.retryAttempts; attempt++) {
      try {
        const url = this.getUrl() + path
        return await this.executeRequest(url, method, body, requestTimeoutMillis)
      } catch (error) {
        lastError = error as Error

        const err = error as Error & { status?: number }
        if (err.status && err.status >= 400 && err.status < 500) {
          throw error
        }

        if (attempt < this.retryAttempts - 1) {
          const delay = this.retryDelayMillis * Math.pow(2, attempt)
          logger.warn('HttpClient.retry', { method, path, attempt: attempt + 1, delay, error: (error as Error).message })
          await new Promise(resolve => setTimeout(resolve, delay))
        }
      }
    }

    logger.error('HttpClient.retry', { method, path, error: 'Max retries exceeded', attempts: this.retryAttempts })
    throw lastError
  }

  private async requestWithFailover(method: string, path: string, body: unknown = null, requestTimeoutMillis: number | null = null, affinityKey: string | null = null): Promise<unknown> {
    if (!this.loadBalancer || !this.enableFailover) {
      return this.requestWithRetry(method, path, body, requestTimeoutMillis)
    }

    const urls = this.loadBalancer.getAllUrls()
    const attemptedUrls = new Set<string>()
    let lastError: Error | null = null

    logger.log('HttpClient.failover', { method, path, totalServers: urls.length, affinityKey })

    for (let i = 0; i < urls.length; i++) {
      const url = this.loadBalancer.getNextUrl(affinityKey)

      if (attemptedUrls.has(url)) {
        continue
      }

      attemptedUrls.add(url)

      try {
        const result = await this.executeRequest(url + path, method, body, requestTimeoutMillis)

        this.loadBalancer.markHealthy(url)

        return result
      } catch (error) {
        lastError = error as Error

        const err = error as Error & { status?: number }
        if (!err.status || err.status >= 500) {
          this.loadBalancer.markUnhealthy(url)
        }

        logger.warn('HttpClient.failover', { url, method, path, error: (error as Error).message })
        console.warn(`Request failed for ${url}: ${method} ${path} - ${(error as Error).message}`)

        if (err.status && err.status >= 400 && err.status < 500) {
          throw error
        }
      }
    }

    logger.error('HttpClient.failover', { method, path, error: 'All servers failed', attempted: attemptedUrls.size })
    throw lastError || new Error('All servers failed')
  }

  private getUrl(): string {
    if (this.loadBalancer) {
      return this.loadBalancer.getNextUrl()
    }
    return this.baseUrl!
  }

  async get(path: string, requestTimeoutMillis: number | null = null, affinityKey: string | null = null): Promise<unknown> {
    return this.requestWithFailover('GET', path, null, requestTimeoutMillis, affinityKey)
  }

  async post(path: string, body: unknown = null, requestTimeoutMillis: number | null = null, affinityKey: string | null = null): Promise<unknown> {
    return this.requestWithFailover('POST', path, body, requestTimeoutMillis, affinityKey)
  }

  async put(path: string, body: unknown = null, requestTimeoutMillis: number | null = null, affinityKey: string | null = null): Promise<unknown> {
    return this.requestWithFailover('PUT', path, body, requestTimeoutMillis, affinityKey)
  }

  async delete(path: string, requestTimeoutMillis: number | null = null, affinityKey: string | null = null): Promise<unknown> {
    return this.requestWithFailover('DELETE', path, null, requestTimeoutMillis, affinityKey)
  }

  getLoadBalancer(): LoadBalancer | null {
    return this.loadBalancer
  }
}

File: clients/client-go/http_client.go
package queen

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

// HttpClient handles HTTP requests with retry logic and load balancing.
type HttpClient struct {
	loadBalancer *LoadBalancer
	config       ClientConfig
	client       *http.Client
}

// NewHttpClient creates a new HTTP client.
func NewHttpClient(config ClientConfig) (*HttpClient, error) {
	// Validate and normalize URLs
	var urls []string
	if config.URL != "" {
		urls = []string{config.URL}
	} else if len(config.URLs) > 0 {
		urls = config.URLs
	} else {
		return nil, fmt.Errorf("at least one URL is required")
	}

	validatedURLs, err := ValidateURLs(urls)
	if err != nil {
		return nil, err
	}

	// Apply defaults
	if config.TimeoutMillis == 0 {
		config.TimeoutMillis = ClientDefaults.TimeoutMillis
	}
	if config.RetryAttempts == 0 {
		config.RetryAttempts = ClientDefaults.RetryAttempts
	}
	if config.RetryDelayMillis == 0 {
		config.RetryDelayMillis = ClientDefaults.RetryDelayMillis
	}
	if config.LoadBalancingStrategy == "" {
		config.LoadBalancingStrategy = ClientDefaults.LoadBalancingStrategy
	}
	if config.AffinityHashRing == 0 {
		config.AffinityHashRing = ClientDefaults.AffinityHashRing
	}
	if config.HealthRetryAfterMillis == 0 {
		config.HealthRetryAfterMillis = ClientDefaults.HealthRetryAfterMillis
	}

	// Create load balancer
	lb := NewLoadBalancer(
		validatedURLs,
		LoadBalancerStrategy(config.LoadBalancingStrategy),
		config.AffinityHashRing,
		config.HealthRetryAfterMillis,
		config.EnableFailover,
	)

	// Create HTTP client with timeout
	httpClient := &http.Client{
		Timeout: time.Duration(config.TimeoutMillis) * time.Millisecond,
	}

	return &HttpClient{
		loadBalancer: lb,
		config:       config,
		client:       httpClient,
	}, nil
}

// Get performs a GET request with retry logic.
func (hc *HttpClient) Get(ctx context.Context, path string, timeoutMs int, affinityKey string) (map[string]interface{}, error) {
	return hc.doRequest(ctx, http.MethodGet, path, nil, timeoutMs, affinityKey)
}

// Post performs a POST request with retry logic.
func (hc *HttpClient) Post(ctx context.Context, path string, body interface{}) (map[string]interface{}, error) {
	return hc.doRequest(ctx, http.MethodPost, path, body, 0, "")
}

// PostWithAffinity performs a POST request with affinity key.
func (hc *HttpClient) PostWithAffinity(ctx context.Context, path string, body interface{}, affinityKey string) (map[string]interface{}, error) {
	return hc.doRequest(ctx, http.MethodPost, path, body, 0, affinityKey)
}

// Delete performs a DELETE request with retry logic.
func (hc *HttpClient) Delete(ctx context.Context, path string) (map[string]interface{}, error) {
	return hc.doRequest(ctx, http.MethodDelete, path, nil, 0, "")
}

// doRequest performs an HTTP request with retry logic.
func (hc *HttpClient) doRequest(ctx context.Context, method, path string, body interface{}, timeoutMs int, affinityKey string) (map[string]interface{}, error) {
	var lastErr error

	// Use custom timeout if provided
	timeout := hc.config.TimeoutMillis
	if timeoutMs > 0 {
		timeout = timeoutMs
	}

	for attempt := 0; attempt <= hc.config.RetryAttempts; attempt++ {
		// Get URL from load balancer
		baseURL := hc.loadBalancer.GetURL(affinityKey)
		if baseURL == "" {
			return nil, fmt.Errorf("no available servers")
		}

		url := baseURL + path

		// Create request body
		var bodyReader io.Reader
		if body != nil {
			jsonBody, err := json.Marshal(body)
			if err != nil {
				return nil, fmt.Errorf("failed to marshal request body: %w", err)
			}
			bodyReader = bytes.NewReader(jsonBody)
		}

		// Create request with context
		reqCtx, cancel := context.WithTimeout(ctx, time.Duration(timeout)*time.Millisecond)
		req, err := http.NewRequestWithContext(reqCtx, method, url, bodyReader)
		if err != nil {
			cancel()
			return nil, fmt.Errorf("failed to create request: %w", err)
		}

		// Set headers
		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("Accept", "application/json")
		if hc.config.BearerToken != "" {
			req.Header.Set("Authorization", "Bearer "+hc.config.BearerToken)
		}
		for key, value := range hc.config.Headers {
			req.Header.Set(key, value)
		}

		logDebug("HttpClient.doRequest", map[string]interface{}{
			"method":  method,
			"url":     url,
			"attempt": attempt,
		})

		// Execute request
		resp, err := hc.client.Do(req)
		cancel()

		if err != nil {
			lastErr = err
			hc.loadBalancer.MarkUnhealthy(baseURL)

			// Check if context was cancelled
			if ctx.Err() != nil {
				return nil, ctx.Err()
			}

			// Check if this is a timeout error (expected for long polling)
			if isTimeoutError(err) {
				logDebug("HttpClient.doRequest", map[string]interface{}{
					"status":  "timeout",
					"attempt": attempt,
				})
				// Don't retry on timeout for long polling
				if timeoutMs > 0 {
					return nil, err
				}
			}

			// Retry with exponential backoff
			if attempt < hc.config.RetryAttempts {
				delay := hc.getRetryDelay(attempt)
				logWarn("HttpClient.doRequest", map[string]interface{}{
					"status":     "retry",
					"attempt":    attempt,
					"error":      err.Error(),
					"retryDelay": delay.Milliseconds(),
				})
				select {
				case <-ctx.Done():
					return nil, ctx.Err()
				case <-time.After(delay):
					continue
				}
			}
			continue
		}

		// Read response body
		respBody, err := io.ReadAll(resp.Body)
		resp.Body.Close()

		if err != nil {
			lastErr = fmt.Errorf("failed to read response body: %w", err)
			continue
		}

		// Check status code
		if resp.StatusCode >= 400 {
			lastErr = &HTTPError{
				StatusCode: resp.StatusCode,
				Body:       string(respBody),
			}

			// Don't retry on 4xx errors
			if resp.StatusCode >= 400 && resp.StatusCode < 500 {
				logError("HttpClient.doRequest", map[string]interface{}{
					"status":     resp.StatusCode,
					"body":       string(respBody),
					"noRetry":    true,
				})
				return nil, lastErr
			}

			// Mark unhealthy on 5xx
			if resp.StatusCode >= 500 {
				hc.loadBalancer.MarkUnhealthy(baseURL)
			}

			// Retry with exponential backoff
			if attempt < hc.config.RetryAttempts {
				delay := hc.getRetryDelay(attempt)
				logWarn("HttpClient.doRequest", map[string]interface{}{
					"status":     resp.StatusCode,
					"attempt":    attempt,
					"retryDelay": delay.Milliseconds(),
				})
				select {
				case <-ctx.Done():
					return nil, ctx.Err()
				case <-time.After(delay):
					continue
				}
			}
			continue
		}

		// Mark healthy on success
		hc.loadBalancer.MarkHealthy(baseURL)

		// Parse response
		var result map[string]interface{}
		if len(respBody) > 0 {
			if err := json.Unmarshal(respBody, &result); err != nil {
				// Try to parse as array
				var arrayResult []interface{}
				if err2 := json.Unmarshal(respBody, &arrayResult); err2 == nil {
					return map[string]interface{}{"data": arrayResult}, nil
				}
				// Return raw body as string
				return map[string]interface{}{"raw": string(respBody)}, nil
			}
		}

		logDebug("HttpClient.doRequest", map[string]interface{}{
			"status":  "success",
			"code":    resp.StatusCode,
		})

		return result, nil
	}

	return nil, fmt.Errorf("request failed after %d attempts: %w", hc.config.RetryAttempts+1, lastErr)
}

// getRetryDelay calculates the retry delay with exponential backoff.
func (hc *HttpClient) getRetryDelay(attempt int) time.Duration {
	delay := hc.config.RetryDelayMillis
	for i := 0; i < attempt; i++ {
		delay *= 2
	}
	return time.Duration(delay) * time.Millisecond
}

// isTimeoutError checks if an error is a timeout error.
func isTimeoutError(err error) bool {
	if err == nil {
		return false
	}
	errStr := strings.ToLower(err.Error())
	return strings.Contains(errStr, "timeout") ||
		strings.Contains(errStr, "timed out") ||
		strings.Contains(errStr, "deadline exceeded")
}

// isNetworkError checks if an error is a network error.
func isNetworkError(err error) bool {
	if err == nil {
		return false
	}
	errStr := strings.ToLower(err.Error())
	return strings.Contains(errStr, "connection refused") ||
		strings.Contains(errStr, "no such host") ||
		strings.Contains(errStr, "network is unreachable") ||
		strings.Contains(errStr, "connection reset")
}

// HTTPError represents an HTTP error response.
type HTTPError struct {
	StatusCode int
	Body       string
}

func (e *HTTPError) Error() string {
	return fmt.Sprintf("HTTP %d: %s", e.StatusCode, e.Body)
}

// GetLoadBalancer returns the load balancer (for testing).
func (hc *HttpClient) GetLoadBalancer() *LoadBalancer {
	return hc.loadBalancer
}

// Close closes the HTTP client.
func (hc *HttpClient) Close() {
	// HTTP client doesn't need explicit closing in Go
	// but we reset the load balancer state
	hc.loadBalancer.Reset()
}
